plugins/fileservices/fileservices.cpp

```diff
@@ -198,63 +198,141 @@ static IPropertyTree *getEnvironmentTree(IConstEnvironment * daliEnv)
     return getHPCCEnvironment();
 }
 
-static StringAttr espurl;           // Default ESP url if none specified
+static void setServerAccess(CClientFileSpray &server, IConstWorkUnit * wu)
+{
+    StringBuffer user, password, token;
+    wu->getSecurityToken(StringBufferAdaptor(token));
+    extractToken(token.str(), wu->queryWuid(), StringBufferAdaptor(user), StringBufferAdaptor(password));
+    server.setUsernameToken(user.str(), password.str(), "");
+}
+
+typedef MapStringTo<time_t> StrMapToEpoch;
+static StrMapToEpoch espUrlsLastHeartBeat;
 static CriticalSection espURLcrit;
 
+static void setWsFSUrlAliveTimeStamp(const char * url, time_t aliveTime)
+{
+    StringBuffer key;
+    CriticalBlock b(espURLcrit);
+    espUrlsLastHeartBeat.setValue(key.set(url).str(), aliveTime);
+}
+
+static time_t contactWsFS(const char * espurl)
+{
+    CClientFileSpray server;
+    server.addServiceUrl(espurl);
+
+    try
+    {
+        Owned<IClientEchoDateTime> req = server.createEchoDateTimeRequest();
+        Owned<IClientEchoDateTimeResponse> result = server.EchoDateTime(req);
+
+        if (result->getResult())
+        {
+            CDateTime now;
+            now.setNow();
+            return now.getSimple();
+        }
+    }
+    catch(...)
+    {
+        PROGLOG("Could not contact WsFS: %s",espurl);
+    }
+    return 0;
+}
+
+static const char * getNextAliveWsFSURL()
+{
+    CriticalBlock b(espURLcrit);
+    CDateTime now;
+    now.setNow();
+    HashIterator iter(espUrlsLastHeartBeat);
+
+    ForEach(iter)
+    {
+        const char * key =  ((const char *)iter.query().getKey());
+        time_t  lastAlive = *(espUrlsLastHeartBeat.getValue(key));
+
+        if (now.getSimple() - lastAlive < 60 * 60)
```

Comment:
Polling intervals are always difficult, the right answer is usually to make it configurable, however based on your overall recommendation, this issue becomes moot.